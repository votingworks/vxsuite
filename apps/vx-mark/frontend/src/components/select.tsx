// Inspiration: https://www.filamentgroup.com/lab/select-css.html
import styled from 'styled-components';

interface Props {
  fullWidth?: boolean;
}

export const Select = styled.select<Props>`
  display: block;
  margin: 0;
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211, 211, 211) url('/images/select-caret.svg') right 0.75em
    top 50%/0.75em auto no-repeat;
  width: ${({ fullWidth }) => (fullWidth ? '100%' : undefined)};
  max-width: 100%;
  padding: 0.75rem 1.75rem 0.75rem 1rem;
  line-height: 1.25;
  appearance: none;
  &:disabled {
    background: #dddddd;
    color: rgb(170, 170, 170);
  }
`;
