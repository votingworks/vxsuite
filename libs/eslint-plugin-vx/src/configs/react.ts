import recommended = require('./recommended');

const jsExtensions = ['.js', '.jsx'];
const tsExtensions = ['.ts', '.tsx'];
const allExtensions = jsExtensions.concat(tsExtensions);

export = {
  ...recommended,
  extends: [
    ...recommended.extends
      .map((name) => (name === 'airbnb-base' ? 'airbnb' : name))
      .filter((name) => name !== 'plugin:prettier/recommended'),
    'airbnb/hooks',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:prettier/recommended', // Enables eslint-plugin-prettier and displays prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
  ],
  settings: {
    ...recommended.settings,
    react: {
      version: 'detect',
    },
  },
  rules: {
    ...recommended.rules,
    'jsx-a11y/label-has-associated-control': [
      'error',
      {
        controlComponents: ['Select'],
      },
    ],
    'react/jsx-filename-extension': ['error', { extensions: allExtensions }],
    'react/jsx-fragments': ['error', 'element'],
    'react/jsx-no-bind': 'off',
    'react/jsx-no-constructed-context-values': 'off',
    'react/jsx-no-useless-fragment': ['error', { allowExpressions: true }],
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-props-no-spreading': 'off',
    'react/jsx-wrap-multilines': 'off',
    'react/prop-types': 'off',
    'react/require-default-props': 'off',
  },
};
