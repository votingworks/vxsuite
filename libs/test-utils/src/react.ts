/* eslint-disable no-console -- this file monkey-patches `console` */

/**
 * Suppresses the "Can't perform a React state update on an unmounted component" warning by
 * modifying the `console.error` method to ignore the warning.
 *
 * This is a temporary workaround until we upgrade to React 18.
 */
export function suppressReact17UnmountedWarning(): void {
  const originalConsoleError = console.error;
  console.error = (...args: unknown[]) => {
    if (
      typeof args[0] === 'string' &&
      args[0].startsWith(
        `Warning: Can't perform a React state update on an unmounted component`
      )
    ) {
      return;
    }

    originalConsoleError(...args);
  };
}
