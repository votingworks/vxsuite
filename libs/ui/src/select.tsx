// Inspiration: https://www.filamentgroup.com/lab/select-css.html
import styled from 'styled-components';

interface Props {
  readonly fullWidth?: boolean;
  readonly large?: boolean;
  readonly small?: boolean;
  readonly primary?: boolean;
}

export const Select = styled.select<Props>`
  display: inline-block;
  margin: 0;
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background-color: ${({ primary }) =>
    (primary && 'rgb(71, 167, 75)') || 'rgb(211, 211, 211)'};
  /* stylelint-disable-next-line string-no-newline */
  background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 287.87"><path style="fill:${({
    primary,
  }) =>
    (primary && 'white') ||
    'darkslategrey'}" d="M502.54,9.46A30.83,30.83,0,0,0,479.78,0H32.22A30.84,30.84,0,0,0,9.63,9.46,30.81,30.81,0,0,0,0,31.87,30.8,30.8,0,0,0,9.46,54.46l224.13,224q22.41,18.91,44.82,0L502.54,54.29q19-22.49-.17-44.83Z" /></svg>');
  background-position: right
    ${({ large, small }) =>
      (small && '0.5em') || (large && '1.125em') || '0.75em'}
    center;
  background-repeat: no-repeat;
  background-size: ${({ large, small }) =>
      (small && '0.5em') || (large && '0.85em') || '0.75em'}
    auto;
  width: ${({ fullWidth }) => (fullWidth ? '100%' : undefined)};
  max-width: 100%;
  padding: ${({ large, small }) =>
    (small && '0.35em 1.25em 0.35em 0.5em') ||
    (large && '1em 2.5em 1em 1.75em') ||
    '0.75rem 1.75rem 0.75rem 1rem'};
  line-height: 1.25;
  color: ${({ primary }) => (primary && '#FFFFFF') || 'black'};
  font-size: ${({ large }) => large && '1.25em'};
  appearance: none;
  &:disabled {
    background: #dddddd;
    color: rgb(170, 170, 170);
  }
`;
