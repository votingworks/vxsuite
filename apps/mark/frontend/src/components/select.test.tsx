import React from 'react';
import { render } from '@testing-library/react';

import { Select } from './select';

const options = (
  <React.Fragment>
    <option value="a">a</option>
    <option value="b">b</option>
    <option value="c">c</option>
  </React.Fragment>
);

it('Inline select', () => {
  const { container } = render(<Select>{options}</Select>);
  expect(container.firstChild).toMatchSnapshot();
});

it('Full-width select', () => {
  const { container } = render(<Select fullWidth>{options}</Select>);
  expect(container.firstChild).toMatchSnapshot();
});
