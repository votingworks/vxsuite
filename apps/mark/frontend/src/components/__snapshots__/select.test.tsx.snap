// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Full-width select 1`] = `
.c0 {
  display: block;
  margin: 0;
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211,211,211) url('/images/select-caret.svg') right 0.75em top 50%/0.75em auto no-repeat;
  width: 100%;
  max-width: 100%;
  padding: 0.75rem 1.75rem 0.75rem 1rem;
  line-height: 1.25;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0:disabled {
  background: #dddddd;
  color: rgb(170,170,170);
}

<select
  class="c0"
>
  <option
    value="a"
  >
    a
  </option>
  <option
    value="b"
  >
    b
  </option>
  <option
    value="c"
  >
    c
  </option>
</select>
`;

exports[`Inline select 1`] = `
.c0 {
  display: block;
  margin: 0;
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211,211,211) url('/images/select-caret.svg') right 0.75em top 50%/0.75em auto no-repeat;
  max-width: 100%;
  padding: 0.75rem 1.75rem 0.75rem 1rem;
  line-height: 1.25;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0:disabled {
  background: #dddddd;
  color: rgb(170,170,170);
}

<select
  class="c0"
>
  <option
    value="a"
  >
    a
  </option>
  <option
    value="b"
  >
    b
  </option>
  <option
    value="c"
  >
    c
  </option>
</select>
`;
